lib/auth/users/usersv1/okta.go

```diff
@@ -0,0 +1,89 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package usersv1
+
+import (
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/authz"
+)
+
+// Okta-origin users have some special access rules that are implemented in this
+// file:
+//
+// 1. Only the Teleport Okta service may create an Okta-origin user.
+// 2. Only the Teleport Okta service may modify an Okta-origin user.
+// 3. Anyone with User RW can delete an Okta-origin user (otherwise there is no
+//    recourse for a Teleport admin to clean up obsolete users if an Okta
+//    integration is deleted.
+//
+// The implementation of these rules is spread through the functions below
+
+// checkOktaOrigin checks that the supplied user has an appropriate origin label
+// set. In this case "appropriate" means having the Okta origin set if and only
+// if the supplied auth context has the build-in Okta role. An auth context
+// without the Okta role may supply any origin value *other than* okta
+// (including nil).
+// Returns an error if the user origin value is "inappropriate".
+func checkOktaOrigin(authzCtx *authz.Context, user types.User) error {
```

Comment:
Looks like functions in this file are duplicated from auth_with_roles.go. Can we reuse?

Reply:
The user management in `auth_with_roles.go` is slated to be deleted in 16.0, and this will go with it, so I didn't spend much time trying to make something that worked for both cases. Happy to try and merge if we think its worthwhile.

Reply:
Actually, done